[RFC] New compiler algorithm

fcard opened the pull request:
# About this

This PR replaces the old transpiler with a more traditional and robust algorithm, with a standard `character stream → tokens → ast → compiled code` logic. This is for the sake of easier development, as the old code was starting to grow too complex and scattered as the corner cases started popping up.

This also throws in a lot of refactoring to fit better with the new compiler code, making use of new functionality that was made easier to implement.

# Functionality

List of things to do.
Feel free to point out anything I might have forgotten.

- [X] Lexer
  - [X] character stream
  - [X] names
  - [X] strings
  - [X] special characters
  - [X] indentation
  - [x] comments
  - [X] handle eof and incomplete inputs
- [x] Ast Parser
  -  [x] expressions
      - [X] names
      - [X] strings
      - [x] variable substitutions
          - [x] simple variable
          - [x] array
          - [x] posix expansions
          - [x] bash expansions 
      - [x] command substitutions
  -  [x] assign
      -  [X] string
      -  [x] numeric array
      -  [x] associative array
  -  [X] commands
  -  [X] blocks and indentation
  -  [X] function definitions
      -  [X] argument list
      -  [X] body
  -  [x] math
      - [X] integer
      - [x] floating point
  -  [x] piping
  -  [x] conditionals
      -  [x] tests
      -  [x] if statement
      -  [x] while statement
      -  [x] switch statement
      -  [x] for statement
  -  [x] powscript extensions
      - [x] async
      - [x] await
        - [x] then
        - [x] pipe
        - [x] lines
      - [x] for of
      - [x] for each line
-  [x] Library
    -  [x] library functions
    -  [x] module system
-  [x] Backends
    -  [x] bash
        - [X] interactive
        - [X] basic functionality
        - [x] all target features covered
        - [x] complete functionality
    -  [x] posix shell
        - [x] interactive
        - [x] basic functionality
        - [x] all target features covered
-  [x] Tools
    - [x] option handling
        - [x] help
        - [X] passing files
        - [x] pipes
        - [X] interactive vs compiler mode
        - [X] pick backend
    - [x] file compiler
    - [x] interactive mode
        - [X] basic functionality
        - [X] handle incomplete input
        - [x] configuration
            - [x] .powrc
            - [x] environment variables (prompt, history settings, etc)
    - [x] tests
        - [x] integrate test system
        - [x] compiler tests
        - [x] passing tests
-  [x] "All files into one" script

# Interactive Mode

One big focus of this PR was to give interactive mode an overhaul, since this more analytic compiler makes it much easier to implement features for it.

![live_powscript_november2-s](https://user-images.githubusercontent.com/12222927/32951220-eaa36480-cb87-11e7-8a48-4b633da825fa.gif)

As a start, it now allows incomplete input (as seen in the gif above) and also separates the host evaluator from the target one. i.e. instead of simply calling `eval` on compiled code, said code is transferred via pipes to a new process and evaluated there, with the result given back to the powscript process. This allows both a separation of namespaces and, more importantly, allows interactive mode to handle other backends, like posix shell or whatever other targets we might have in the future.

**Old**
![old_interactive_mode - page 1](https://user-images.githubusercontent.com/12222927/30144088-7c76e00a-935f-11e7-9c0a-2eeac69ba86c.png)

**New**
![interactive_mode_diagrams - page 1](https://user-images.githubusercontent.com/12222927/30144202-1915a9b4-9360-11e7-8e08-6da5e72ae5d9.png)

# Style

This new approach also needed a new style to make it work more naturally and efficiently.

## Subshells and `echo` → Passing names and `setvar`

I've tried to avoid subshells as much as possible, not only are they slow but also make the code more complex and harder to reason about, and make abstracting global state near impossible.

Instead of:
```sh
f() {
  local value="$1"
  echo "returning $value"
}
Var="$(f argument)"
```
I used:
```sh
f() { #<<NOSHADOW>>
  local value="$1" out="$2"
  setvar "$out" "returning $value"
}
noshadow f 1

f argument Var
```
Where `setvar name value` sets the variable `name` with the value `value`, and `noshadow fun argnum=0 outnum=1` restructures the function `fun` so the given variable names are not shadowed by local variables. `#<<SHADOW>>` makes `.tools/compile` apply the `noshadow` function so only the result is present in the final powscript file. (making startup faster)

While at first glance a bit more verbose and unorthodox compared to echoing the results, it:
1. Removes dynamic scope, making the code easier to understand.
2. Allows multiple return values, easily arranged.
3. Allows better abstractions.
4. Frees output for other purposes more suited for it.
5. Separates return values from other types of printing.
6. More suited to the style required by compiler code. (manipulating general data instead of just strings)
7. Faster.

It also becomes comparatively less verbose and messy than `echo` et al the larger the function gets and the more abstract we become.
```sh
ast:parse ast
ast:from $ast value ast_value
ast:from $ast head  ast_head

# vs

parse_ast # can't return value as subshell would not allow modification to token stack
ast=$(last_ast)
ast_value="$(from_ast $ast value)"
ast_head="$(from_ast $ast head)"
```
```sh
token:get -v value -c class

# vs

get_token
token_id=$(last_token)
value="$(from_token $token_id value)"
class="$(from_token $token_id class)"
```

This is similar to the way the `read` function operates and is implemented on top of `printf -v`.

## Abstractions

The main form of abstraction is that of an associative array mapping numeric ids to structures, allowing a more organized approach to manipulating data.
```sh
ast:make ast name
ast:from $ast head #> name
```
Some of the structures also have a stack-like quality keeping the last created element accessible and allowing backtracking.
```sh
{
  stream:init

  token:get -v value
  echo $value #> hello

  token:get -v value
  echo $value #> world

  token:backtrack
  token:backtrack

  token:get -v value
  echo $value #> hello

} <<< "hello world"

```

**Basic format:**

|   Key     | Value |               Description                   |
|-----------|---------|------------------------------------------|
| index    |     7     |  Id of last instance created + 1   |
| 0-field1 |     x     |  field1 of instance of id 0            |
| 0-field2 |     y     |  field2 of instance of id 0            |
| 1-field1 |     a     |  field1 of instance of id 1            |
| 1-field2 |     b     |  field2 of instance of id 1            |
|   ...        |    ...     |  ...                                              |

# Structure

A few refactoring have been done to file and directory structure to allow easier development.

## File organization

Files are organized in a more modular way. Compiler related files are all placed in the *src* directory.
- **src/lexer/**: related to transforming a stream of characters into tokens.
- **src/ast/**: transform tokens into tree-like structures ready to be transformed into compiled code
- **src/lang/**: contains functions specific to each backend (for now just bash and sh, but structured to make it easier to add more in the future)
- **src/compiler/'**: brings all the previous code together to make the compiler tools
- **src/powscript.bash**: entry point

Library files have been moved to *lib*, and will eventually be glued into a `powscript_std.pow` or somesuch where they will be subject to a generalized module system. That way, we can expand the "conditional including" of functions to user modules.

Tests are in the process of being reorganized as well, all files used by the tests are in *test/test-files* instead of alongside them, and a few loose tests have been put into specific categories.

## New "Compile all files into one" strategy

A new, more dynamic method of compiling the sources into a single `powscript` file was devised <s>(but isn't implemented yet)</s>:

```sh
powscript_source "ast/ast.bash"  #<<EXPAND>>
```
The directive `#<<EXPAND>>` above will cause the compiler script to remove this line and add the contents of the `ast/ast.bash` file in its place. This allows the files to be tested on their own while at the same time making it easier to add or remove files without having to change a hard-coded script.

---

fcard committed:
More traditional compiler algorithms (WIP)

fcard committed:
usable tokenizer

fcard committed:
Something is working!

fcard commented:
Welp, that's a lot of changes, but hopefully I made a good case for them. 

This still needs quite a bit of work but I felt it would be better to present it as soon as I had something running + a complete outline of the goal, in case you have anything to add or criticize.

fcard commented:
Status update: Still working on this, mostly fixing bugs (I will need to update that gif, having to add two newlines to end a block wasn't supposed to be necessary and is now fixed)

I want to have this done by the end of the year, but can't guarantee due to college + another project I need to focus on. I will at least promise that unless I suffer a bus crash I will not leave this unfinished!

I vow on my honor as a coder! As an aspiring professional! As someone people can rely on! This. **Will.** ***Be.*** ***DONE!***

---

<sub><sup>Oof</sup></sub>
Got a little carried away there, but I mean it :)

coderofsalvation commented:
hehe thx :) no worries. Looking good!

fcard committed:
bugfixes and more things working:
  - function arguments
  - variable and command substitution
  - if/elif/else blocks

fcard committed:
bugfixes + multiline command substitution

fcard committed:
numeric array creation and indexing

fcard committed:
.tools/compile now working with new code.
$var[i] implemented as equivalent to ${var[i]}

fcard committed:
simplifications of the ast parsing algorithms
+ nested block and other bugfixes

fcard committed:
better whitespace in noshadow output

fcard committed:
start updating tests

fcard committed:
bugfixes:
- Kill guest process on exit
- Check for pipes before deletion
- Fixed flags in get_token

fcard committed:
implement for loop

coderofsalvation commented:
still looking good! Just wanted to tell you I'm really excited about all changes so far.

fcard committed:
consistent naming convention.
+ Weird makeshift regex debugger. Better than adding echoes everywhere.+ bugfixes

fcard committed:
bugfixes related to assigning values

fcard commented:
Glad to hear it! I am almost done implementing all the bash stuff, then I will move on to the powscript extensions and library. I'd like to hear your opinion on #41 and #42 since they can influence the way I go about this.

fcard committed:
Add posix shell compilation
Now bash compilation is implemented in terms of sh.

fcard committed:
Associative arrays definition syntax (x={})
+ square bracket arrays (x=[])
+ variable substitution bugfixes

fcard committed:
Implement math
+ implement x+=n, x-=n,...
+ x@=y is the push assign
+ $#x gives the array length of x

fcard committed:
bugfix: ignore empty ASTs

fcard committed:
Implement flag conditionals
+ make ast:print work with conditionals, for loops, math and array length/push

fcard commented:
After implementing math (as a special op so `echo $(math 1+2)` becomes `echo $((1+2))`, bypassing the command substitution), I added the assignment shorthands as a bonus

![`x+=1`](https://user-images.githubusercontent.com/12222927/33053895-d98fee50-ce54-11e7-86a1-38c6f1bbcf0e.gif)
(works also for `x-=n`, `x*=n`, `x/=n`, `x^=n`, `x%=n`)

---

But then I remembered that `x+=y` is meant to be the push operation, so I implemented it as `x@=y` instead.

![`x@=1`](https://user-images.githubusercontent.com/12222927/33053984-4f98c36a-ce55-11e7-960d-6339d1601ac0.gif)

Is that alright? Any suggestions?

---

While we're at it, I am trying not to add new features and leaving that for after this PR otherwise this would take even longer, but some were needed along the way or were exceptionally easy to do. So far, this adds:

- Support for `elif` in `if` statements.
- Flag tests in `if` statements. (e.g. `if -x powscript and -d src`)
- `^` for exponentiation instead of `**` to keep it consistent with `bc`.
- `$#Array` gets the array's length.
- `x=[1 2 3 4]` is the same as `x=(1 2 3 4)`.
- On POSIX mode, terminates compilation if bash features are found. (reporting which ones)
- POSIX interactive mode.
- Debug tools on interactive mode:
  * `.compile` displays the compiled code after every command. (toggle)
  * `.ast` displays the AST after every command. (toggle)
  * `.echo` displays every command as they were passed to the compiler. (toggle)
  * `.tokens <code>` lists the tokens of the given code.
  * `.show <ast>` inspects the AST with the given ID.
- gdb-like debugger for the compiler (`POWSCRIPT_CDEBUG=true .tools/compile`), which is a ridiculous hack but it's been working pretty well so far. Commands:
  * `break <function name>` adds breakpoint at the start of the given function.
  * `step` move forward, entering function calls.
  * `next` move forward, jumping over function calls.
  * `end` move to the end of the current function call.
  * `skip n` skip n lines.
  * `continue` proceed with execution until next breakpoint is found.
  * `test <code>` evaluates the given code and prints `true` or `false` depending on its return value.
  * `var <variable 1> <variable 2>...` displays the values of the given variables.
  * `eval <code>` evaluates code in the current context.
  * `line` show the line we're currently stopped at.
  * `quit` exit debugger.

If there is anything you dislike please tell me so I can change it.

fcard committed:
bugfix: command substitution didnt terminate
+ githook now checks if compiler is in debug mode

fcard committed:
Implement == as numeric equality
+ exit on unimplemented feature when in non-interactive mode
+ simplify compilation code a bit

fcard committed:
Divide ast parser into more files, add more comments

fcard committed:
implement switch statement
+ bugfixes: indent error, numeric comparison in POSIX mode

fcard committed:
Make switch/case AST printable, separate printing in its own file
+ Refine error message in trailling expression after switch/case

fcard committed:
Implement floating point math
+ math bugfixes with parens

fcard committed:
While statement
+ bugfix: seperate pattern parsing from expression parsing

coderofsalvation commented:
I definately like everything so far.
I'm thinking about the += thingie:

* do we want it as a numeric operation
* do we want to keep it as an native bash array push operation
* does `x@=1` occur in other languages? (php does `x[]=1`, and javascript does `x.push(1)` for example)

fcard commented:
Out of all the possibilities, the numeric one is the one that saves most characters:
```sh
x=$(math x+1)
x=$x'1'
x[$#x]=1
```
Plus it goes with the other shorthands `x-=n`, `x*=n`...

However, since strings are the primary element in shells, there is a strong case for it to mean string concatenation, in which case I should get rid of the other shorthands.

(Note that `+=` in bash is string concatenation, not array push:

```sh
x=()
x+=1
[ "${x[0]}" = 1 ] # looks like array push, but...
x+=2
[ "${x[1]}" = "" ] # oops!
[ "${x[0]}" = 12 ] # it's just concatenating to the first value
```
**edit**: Actually it looks like bash overloads the operator and if you declare a variable as an integer `+=` becomes a numeric operation for it.
)

---

I can't think of any language that uses `@=`, it invokes arrays in bash (and powscript) but there are probably better alternatives. `x[]=y` could work but I don't see it being clearer, it looks like an assignment to an associative array with an empty key.

`x.push` is an OOP thing, and would be misleading unless we implement other methods like `x.length` or `x.pop`. Seems out of place either way. It could just be a function/macro, `push x 1`.

fcard commented:
Personally, I like keeping math operators only doing math things, so I would keep the numeric assignments and find another symbol for string concatenation. [(a lot of languages seem to like `&`)](https://www.rosettacode.org/wiki/String_concatenation)

Another advantage of not just using the default bash `+=` and writing our own instead is that we can make it work in POSIX mode. Also the more control we have over our constructs the better.

Since we apparently never actually had a `push` operator maybe we can leave it for after this PR. We can open a separate issue and have the discussion there (not for `+=` & co though because I need to determine the fate of the concatenation operator to finish this, as I don't want us to lose any features)

coderofsalvation commented:
agreed, in essence they're math operators.
Thank you for this breakdown & your thoughts, we're on the same page about this.
The goal is to make things simple and straightforward.

fcard commented:
I've been having some technical problems so my goal of finishing this by the end of the year seems like it's not gonna happen :/ But just as a status update I have implemented all operators including concatenation, plus a bunch of other stuff which should conclude the Bash part of this, but I still need the tests to make sure.

For now I am leaving `&=` as concatenation and `@=` as array pushing, we can easily change it later if need be.

I decided to add a "lowering" step to the compiler to simplify some of the AST parsing functions, I need to reorganize some stuff but it shouldn't take long.

The discovery of the context sensitivity of the `+=`  operator in Bash got me thinking about its type system and how powscript doesn't have one. It's something worth of discussion imo since type information opens many doors for optimizations and smarter compilation. There is also no target agnostic way to define a integer in powscript:
```sh
declare -i count
```
will compile but not work in POSIX mode. I can just make the `declare` command throw a compilation error in POSIX mode but I always thought it would be neat to eventually have all powscript features available in it, e.g. we could make
```sh
X={}
X[a]=1
echo $X[a]
```
compile to
```
__powscript_assoc_X=true
__powscript_assoc_X_a=1
echo "$__powscript_assoc_X_a"
```
Or whatever. Having a type system would be a step towards that.

The details would need to be worked out later, for now I am just making the lowering step tag every variable and expression as `?` typed.

fcard committed:
add lowerer step, implement our own declare to add type information
+ simplications and optimizations

fcard commented:
That's that all done.
I special-cased the declare command so it gives us type information:
![live_powscript_declare-s01](https://user-images.githubusercontent.com/12222927/34468169-ffe21bfc-eee1-11e7-9373-fae877c8406c.gif)

Accepted types are `integer`, `string`, `array` and `map`. In POSIX mode, `declare` now just becomes `local`, so it's now portable to use it.

fcard commented:
I am now implementing `for of`. I thought it was for both numeric and associative arrays, but looking at your code it's only for associative ones.

@coderofsalvation, what do you think of using `for x of Array` for numeric arrays? The current behavior where `for x in Array` is turned into `for x in $Array[@]` but `for x in a b c` remains a normal bash `for` loop seems a bit too magical.

coderofsalvation commented:
Hi fcard, thx for all the awesome progress so far!

How would you imagine those for loops?
Right now we have `for x in a b c` (posix) and for associative arrays:

```
foo={}
foo["bar"]="a value"
for k,v in foo
echo k=$k
echo v=$v
```

How would you envision a revised version?

fcard commented:
Hmm, I didn't realize `in` also worked for associative arrays. I was using `code-1.pow` and the other tests for reference and all uses of `for` for associative arrays use `of`.

I've made this work already:
```sh
A={}
A[key]=value
for k,v of A
  echo k=$k v=$v
```

And this as well actually:
```sh
A=(1 2 3)
for v of A
  echo $v
```

And I am fine with keeping `in` for normal shell `for`s and `of` for powscript's special `for`s, it's simpler than "`for <x> in <y>` *might* be special depending on `<x>` and/or `<y>`". We could make it more explicit by using `foreach`:
```sh
Anum=(1 2 3)
Assoc={}
Assoc[key]=value

foreach x in Anum
  echo $x

foreach k,v in Assoc
  echo k=$k v=$v
```
(which then means `foreach` could be implemented as a macro...)

Either is fine by me. I can also make `in` as it is now, it's your call.

coderofsalvation commented:
ah ok, how about just leaving it `for ... of` then.

Then the only thing we need to update is line 115, 129, 150, 164 in reference.md

fcard committed:
for-of

fcard commented:
That's that done. I haven't updated *reference.md* yet but I will once I have all the features finished so I can do it all in one go.

 I am already overdue so I think I will save the macros/value substitution for future PRs and implement everything as special ops for now. Not ideal but it shouldn't be too hard to refactor this.

Now all I need to do is:
* `await then` and `when done`
*  simple module system
* library functions
* command line utilities
* basic config files groundwork (*.powrc*, some config environment variables)
* tests

We're about 95% done here I think. I would like to give a set date but having a mix of hardware problems and college getting in the way makes that hard. But hopefully I can finish this soon so I can start working on other features.

fcard committed:
add await (then and pipe)
+ minor bugfixes in ast:print

fcard commented:
Back to this. I took much longer to return than I expected, because I was hit by the metaphorical bus and fell on the literal floor, but here I am. Now we have `await ... then` and `await ... then |`.

![live_powscript_await-s0](https://user-images.githubusercontent.com/12222927/36939292-378a5afa-1f0d-11e8-8812-a1d336f8f0a4.gif)

coderofsalvation commented:
awesome!
this will make writing bash much more readable.

On Sat, Mar 3, 2018 at 10:09 PM, fcard <notifications@github.com> wrote:

> Back to this. I took much longer to return than I expected, because I was
> hit by the metaphorical bus and fell on the literal floor, but here I am.
> Now we have await ... then and await ... then |.
>
> [image: live_powscript_await-s0]
> <https://user-images.githubusercontent.com/12222927/36939292-378a5afa-1f0d-11e8-8812-a1d336f8f0a4.gif>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coderofsalvation/powscript/pull/38#issuecomment-370180037>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAK_ZHvfwm-Z4oc6m2B6mmtg54c7567Xks5tawadgaJpZM4PPQIf>
> .
>



-- 
/WEBSITE    2wa.isvery.ninja
/MOBILE     +31 6 13499604
/LOCATION   Netherlands, Drachten. Hongarije, Budapest.
/COMPANYNR. 08124656
/TAGS       Rapid Prototyping of ideas/concepts using JavaScript,
            Node.js, OpenSource. Growth-hacking partnerships,
            lead scraping partnerships.

fcard committed:
for-from + await-for

fcard commented:
`for <line> from <file>` is here, as is `await <command> then for <line>`.

![live_powscript_foreach_line-su2](https://user-images.githubusercontent.com/12222927/36979179-72c697ba-2065-11e8-941a-21d0cce318d9.gif)
![live_powscript_await_for-s2](https://user-images.githubusercontent.com/12222927/36979197-8015835e-2065-11e8-9bdb-ed1284e68459.gif)

Now the only things left are the library functions, finishing the documentation and reintegrating tests.

coderofsalvation commented:
oh somehow my reply didn't get through, here goes again:

this is very awesome.
Imho the await makes things so much more readable.

fcard commented:
I am back! I am really sorry, college is much more demanding than I expected it to be.

While implementing the library functions I run into a problem: I wanted to make the implementations more backend-independent, and to do so by implementing them as macros, e.g. `map` would be:

```julia
macro map(f A)
  code
    for x of ~A
      ~f $x
```
(or an equivalent to this, since I am not implementing macros just yet)
It would only use powscript, without needing one implementing a `foreach` for bash, another for `sh`, etc... However, then the functions become less flexible for when they were to be passed as arguments.
```julia
macro twice_macro(f A)
  code
    ~f ~A
    ~f ~A

twice_function(f A)
  $f $A
  $f $A
  
twice_macro map A      # works if map is either a macro or a function
twice_function map A   # works only if map is a function
```
So instead of that, I was thinking of using the following compromise:
```julia
map(f A)
  expand
    for x of ~A
      ~f $x
```

Where, on bash and sh, `expand powscript-code` will become `eval "backend-code"` with the `$` and strings etc escaped and `~` replaced by `$`. e.g. on `bash` the above would become:

```sh
map() {
  local f="$1"
  local A="$2"
  eval "
    for x in \"\${$A[@]}\"; do
      echo \"\${x}\"
    done
  "
}
```

What do you think?

coderofsalvation commented:
hey no worries! life can be very busy, i know all about that :)

I think the eval solution works out very elegantly. Thanks for adding this
backend-agnostic layer.
It really makes thinks more maintainable/readible.
Thx!!Huge milestone!


On Sun, May 6, 2018, 19:57 fcard <notifications@github.com> wrote:

> I am back! I am really sorry, college is much more demanding than I
> expected it to be.
>
> While implementing the library functions I run into a problem: I wanted to
> make the implementations more backend-independent, and to do so by
> implementing them as macros, e.g. map would be:
>
> macro map(f A)
>   code
>     for x on ~A
>       ~f $x
>
> (or an equivalent to this, since I am not implementing macros just yet)
> It would only use powscript, without needing one implementing a foreach
> for bash, another for sh, etc... However, then the functions become less
> flexible for when they were to be passed as arguments.
>
> macro twice_macro(f A)
>   code
>     ~f ~A
>     ~f ~A
> twice_function(f A)
>   $f $A
>   $f $A
>
> twice_macro map A      # works if map is either a macro or a function
> twice_function map A   # works only if map is a function
>
> So instead of that, I was thinking of using the following compromise:
>
> map(f A)
>   expand
>     for x on ~A
>       ~f $x
>
> Where, on bash and sh, expand powscript-code will become eval
> "backend-code" with the $ and strings etc escaped and ~ replaced by $.
> e.g. on bash the above would become:
>
> map() {
>   local f="$1"
>   local A="$2"
>   eval "    for x in \"\${$A[@]}\"; do      echo \"\${x}\"    done  "
> }
>
> What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coderofsalvation/powscript/pull/38#issuecomment-386898654>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAK_ZIoJpL5-SZb5OZLl8LRp3VbKqLBeks5tvzlrgaJpZM4PPQIf>
> .
>

fcard commented:
Since I am close by, here's a status update:
![mapc](https://user-images.githubusercontent.com/12222927/40800701-7b241cea-64e7-11e8-9d38-3447888b66fe.gif)

fcard committed:
expand: portable eval

fcard committed:
Add standard library + command substitution fix

fcard commented:
The standard library has been implemented! I went ahead and added `require_env` and `first` as well. Can you take a look at the [*std.pow* file](https://github.com/fcard/powscript/blob/fa0668e7a2a1f733399d09ba4cf9a04b4c598165/lib/std.pow) and confirm that everything is as it should?

The `json_decode` function isn't there yet, but I am pretty sure I am just going to copy yours, since it's pretty large. We can port it to powscript later.

What is [`on`](https://github.com/coderofsalvation/powscript/issues/2) meant to do?

fcard committed:
Add help texts.

fcard committed:
add support for .powrc file

fcard committed:
require

fcard committed:
Reintroduce and get all tests passing

fcard committed:
More help text

fcard commented:
This is finished, pretty much. All functionality is in, I just need to test for bugs, add documentation, and improve performance in some places.

fcard committed:
Treat unsupported variable expansions as errors

fcard committed:
Start fixing reference.md
+ add escaped newline support in interactive mode

fcard committed:
More corrections to reference.md irt #38

fcard committed:
Make flag conditions work again

fcard committed:
add string manipulation operators.
+ fix math special op when in command substitution

fcard commented:
I added a few more features.

`assert`: Exits with status 1 if the condition is false.
```sh
x=1
assert $x is 1 # does nothing
assert $x is 2 # exits
```
String indexing operators (all but length are bash only for now):
```sh
x=abc
echo ${x:length}           # 3
echo ${x:index 0}          # a
echo ${x:from 1 to 2}      # bc
echo ${x:slice 0 length 2} # ab
```

String manipulation operators (renamed to be easier to remember, plus a few tweaks):
```sh
x=aabbcc
echo ${x:prefix  a*}       # abbcc (strips whitespace surrounding expressions)
echo ${x:prefix* a*}       # bbcc
echo ${x:suffix  *c}       # aabbc
echo ${x:suffix* *c}       # aabb
echo ${x:replace  a by b}  # babbcc (replace is currently bash only)
echo ${x:replace* a by b}  # bbbbcc
echo ${x:replace *a by b}  # bbbcc

y='//'
echo ${y:replace* / by a} # aa (auto-escapes '/'s)
```

String expansion operators (again, renamed):
```sh
echo ${var:unset value}    # (replaces if variable is unset)
echo ${var:empty value}    # (replaces if variable is unset or set to '')
echo ${var:unset= value}   # (replaces and sets if variable is unset)
echo ${var:empty= value}   # (replaces and sets if variable is unset or set to '')
echo ${var:set value}      # (replaces if variable is set)
echo ${var:nonempty value} # (replaces if variable is set to something other than '')
```

Criticism is welcome, if you have any.
I think I changed my mind, and will make an original json_decode function as a standard powscript module. Sounds like a good opportunity to test the compiler for bugs.

fcard committed:
add collection operators
+ fixes bug where \\ was needed to have a  in a pattern
+uppercase/lowercase and set!/nonempty! operators

fcard committed:
Indirect expansion
+ Added ast print support

fcard committed:
Empty expansion: ${}
+ Got tests halting on failure again and several bug fixes

fcard committed:
Fixes bug that single quoted strings in expansion arguments acted wonky

fcard commented:
Recreating the json function turned out to be a good idea, I fixed many bugs and added a few features that making an actual program made me realize would be handy. (all syntax is easily changeable if you don't like them)

Keyword arguments:
```sh
f(a b -- x y)
  echo "a=$a b=$b x=$x y=$y"

f 1 2         #> a=1 b=2 x= y=
f 1 2 --x 3   #> a=1 b=2 x=3 y=
f 1 2 --y 3   #> a=1 b=2 x= y=3
f --x 1 --y 2 #> a= b= x=1 y=2

g(-- argument)
  echo $argument

g --argument 1 #> 1
g -a 1         #> 1
```

Referential arguments: A better solution to the noshadow issue
```sh
f(x)
  x:ref= b

x=a
f ${x:ref}
echo $x #> b

# Solving the following issue:
g(x)
  eval "$x=b"

x=a
g x
echo $x #> a, because the local x in g shadows the passed x

g ${x:ref}
echo $x #> b. x:ref= is a special syntax that makes sure that x is a reference
```

Conditional expansions:
```sh
test ${x:set?}
test ${x:unset?}
test ${x:empty?}
test ${x:nonempty?}
```

Conditional assignments:
```sh
x:set= a
x:unset= a
x:empty= a
x:nonempty= a
```

Unicode:
```sh
echo "\x61" #> a
echo "\u62" #> b
echo "\U63" #> c

echo "\xf0\x9f\x90\xb1" #> 🐱 
echo "\uD83D\uDC31"     #> 🐱 
echo "\U0001F431"       #> 🐱 

require unicode

parse_utf8 "f0 9f 90 b1" -r ${a:ref} -d ' '
parse_utf16 "D83D DC31"  -r ${b:ref} -d ' '
parse_utf32 "0001F431"   -r ${c:ref} -s ''
echo $a $b $c #> 🐱 🐱 🐱
```

And of course, JSON:
```sh
x={}
json_parse x '{"a": 1, "b":2}'

json_print x a #> 1
json_print x b #> 2

json_value ${xa:ref} x a
echo xa #> 1
```
I can implement the old `json_decode` on top of those features easily, although a problem I noticed with it is that object keys cannot have a "-" character in them.

I haven't received any messages from you, by the way, in case they didn't go through like last time.

fcard committed:
JSON, unicode, keywords, references

fcard committed:
Caching

fcard commented:
This is finished. I am still itching to add more stuff but this PR is way too large as it is, and all the initial goals have been met.

coderofsalvation commented:
awesome,
I will do a test, and then I'll merge+publish it.

Leon

On Sat, Jun 9, 2018 at 6:52 AM fcard <notifications@github.com> wrote:

> This is finished. I am still itching to add more stuff but this PR is way
> too large as it is, and all the initial goals have been met.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coderofsalvation/powscript/pull/38#issuecomment-395940049>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAK_ZNGPjwsGxlo9awOjdf-2aPeVk293ks5t61SbgaJpZM4PPQIf>
> .
>


-- 
/WEBSITE    2wa.isvery.ninja
/MOBILE     +31 6 13499604
/LOCATION   Netherlands, Drachten. Hongarije, Budapest.
/COMPANYNR. 08124656
/TAGS       Web & New Media consultancy driven by Opensource,
            Google tech & a teaspoon of javascript.

fcard committed:
Cache progress messages and automatic version updating

fcard commented:
You might have noticed that the startup time is very long the first time you run the new powscript, the reason is it's compiling all the libraries into cache files so it doesn't have to do it every time it's run.

My latest commit added a message specifying this so it's not just a mysterious one-time delay.

Oh, and I also added an automatic version updating bit to the compile script, so that it adds to the patch version every time it notices a change.

coderofsalvation commented:
Excellent, im busy going thru everything.
Notes so far:

* very nice!
* all tests run fine
* startup time isn't an issue as long as its only once
* im reviewing all tests, and im seeing how it affects my old powscripts
* after that i can create a small migration guide, and update the docs

I've created a link to your master- or token-ast-compiler-branch in the README.md, so powscript-users can already try the beta.
In the meantime i will try to create some small PR's to you, with small fixes.
When we're satisfied we'll remove the beta-link, rename the master to 'v0.5', and make token-ast-compiler-branch the new master.

Would ♥♥♥ to hear your feedback on this. 

Leon

fcard commented:
Sounds good to me! Feel free to send any changes or suggestions my way. If you need anything else I'd be glad to help.

coderofsalvation commented:
BOOM! MERGED! 
the v1 BETA landed.  🎉 🏆 🎉 🏆 🎉 🏆 🎉 🏆 🎉 🏆
I've tagged the old powscript as a v0.9-release.
Currently I'm trying to modify the README.md example, to get the powscript snippet working.

Btw. are there files we can get rid of?
Suspects:

* maybe some files in .tools directory?
* lang/bash directory?

> which can we delete?

Oh and:

* how about changing `set?` `unset?` `empty?` to `set` `unset` and `empty` ? Or is there a deeper meaning behind the ? (reason: my old powscripts use `empty` instead of `empty?`)
* I'd like to add `require_cmd` and `require_env` again..where should i put this?

Keep in touch!

ps1. again thanks a lot for your contribution
ps2. Don't forget, you're a powscript contributor so you have push-access.
ps3. Im going to start a new project with powscript v1, so this will keep me more involved in powscript again (my latest project unfortunately pushed my focus away from powscript too much, sorry for that).

fcard commented:
Awesome!! I am really happy you've found my contribution useful!

As for your points:
* I don't see any use for the lang/bash directory anymore, I probably just forgot to delete it :P I don't see any extraneous files in the .tools directory, which ones do you mean?
* The reason I used `set?`, `unset?` and `empty?` is because `set` and `unset` are already commands, and since I put the `?` in two of the predicates, it would be inconsistent to not have it in the last one.
* `require_env` is there in the `lib/std.pow` file, you can add `require_cmd` after it!

Thanks a lot, I look forward to contributing more! As soon as I have time I will port my PRs from my repo to here ;P
